util/conditions/patch.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conditions
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+)
+
+// PatchOperation defines patch operation types.
+type PatchOperation string
+
+const (
+	// AddConditionPatch defines an add condition patch operation.
+	AddConditionPatch PatchOperation = "Add"
+
+	// ChangeConditionPatch defines an change condition patch operation.
+	ChangeConditionPatch PatchOperation = "Change"
+
+	// RemoveConditionPatch defines a remove condition patch operation.
+	RemoveConditionPatch PatchOperation = "Remove"
+)
+
```

Comment:
Can we follow the controller runtime pattern for patches https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/patch.go#L86-L89? For example we could have `conditions.MergeFrom(obj conditions.Getter) Patch` which is already familiar.

Reply:
I will try to use the same name for consistency, but we are solving a slightly different problem than controller runtime because controller runtime does not account for concurrency, instead, we do.

controller runtime compares base & target and computes the differences; instead, we are considering concurrency, so we should consider also current base (or source) that id the object modified by another controller